File: Python/OneLang/One/Ast/AstHelper.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.AstTypes as astTypes

class AstHelper:
    def __init__(self):
        pass
    
    @classmethod
    def collect_all_base_interfaces(cls, intf):
        result = dict()
        to_be_processed = [intf]
        
        while len(to_be_processed) > 0:
            curr = to_be_processed.pop()
            result[curr] = None
            
            if isinstance(curr, types.Class) and curr.base_class != None:
                to_be_processed.append((curr.base_class).decl)
            
            for base_intf in curr.base_interfaces:
                to_be_processed.append((base_intf).decl)
        
        return Array.from_(result.keys())

File: PHP/src/OneLang/One/Ast/AstHelper.php
<?php

namespace OneLang\One\Ast\AstHelper;

use OneLang\One\Ast\Types\IInterface;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\ClassType;

class AstHelper {
    static function collectAllBaseInterfaces($intf) {
        $result = new \OneCore\Set();
        $toBeProcessed = array($intf);
        
        while (count($toBeProcessed) > 0) {
            $curr = array_pop($toBeProcessed);
            $result->add($curr);
            
            if ($curr instanceof Class_ && $curr->baseClass !== null)
                $toBeProcessed[] = ($curr->baseClass)->decl;
            
            foreach ($curr->baseInterfaces as $baseIntf)
                $toBeProcessed[] = ($baseIntf)->decl;
        }
        
        return \OneCore\Array_::from($result->values());
    }
}
